cmd/options/options_test.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package options
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"testing"
+)
+
+type Options struct {
+	Nodename         string
+	Hostname         string
+	HostnameOverride string
+}
+
+// TestSetNodeNameOrDie tests for permutations of nodename, hostname and hostnameoverride
+func TestSetNodeNameOrDie(t *testing.T) {
+	options := map[string]struct {
+		Expected         Options
+		ObtainedNodeName string
+	}{
+		"Check Node and HostnameOverride only": {
+			Expected: Options{
+				Nodename:         "my-node-name",
+				Hostname:         "",
+				HostnameOverride: "override",
+			},
+		},
+		"Check Nodename only": {
+			Expected: Options{
+				Nodename:         "my-node-name",
+				Hostname:         "",
+				HostnameOverride: "",
+			},
+		},
+
+		"Check HostnameOverride only": {
+			Expected: Options{
+				Nodename:         "",
+				Hostname:         "",
+				HostnameOverride: "override",
+			},
+		},
+		"Check Hostname only": {
+			Expected: Options{
+				Nodename:         "",
+				Hostname:         "my-host-name",
+				HostnameOverride: "",
+			},
+		},
+		"Check Node, host and HostnameOverride only": {
+			Expected: Options{
+				Nodename:         "my-node-name",
+				Hostname:         "my-host-name",
+				HostnameOverride: "override",
+			},
+		},
+
+		"Check Host and HostnameOverride only": {
+			Expected: Options{
+				Nodename:         "",
+				Hostname:         "my-host-name",
+				HostnameOverride: "override",
+			},
+		},
+
+		"Check Node and hostname": {
+			Expected: Options{
+				Nodename:         "my-node-name",
+				Hostname:         "my-host-name",
+				HostnameOverride: "",
+			},
+		},
+	}
+
+	orig_node_name := os.Getenv("NODE_NAME")
+	orig_host_name, err := os.Hostname()
+	if err != nil {
+		fmt.Println("Unable to get hostname")
+	}
+
+	for str, opt := range options {
+		if opt.Expected.Nodename != "" {
+			err = os.Setenv("NODE_NAME", opt.Expected.Nodename)
+			if err != nil {
+				t.Errorf("Unable to set env NODE_NAME")
+			}
+		}
+
+		if opt.Expected.Hostname != "" {
+			// Changing hostname
+			cmd := exec.Command("hostname", opt.Expected.Hostname)
```

Comment:
We should avoid setting `hostname` directly. However, setting an environment is ok to me. But you need to clear the `NODE_NAME` env for each test as the in the for loop the prior setted `NODE_NAME` environment is not cleared and will have an impact on the next test case.

Reply:
The more proper way to test hostname is we should first get the hostname and check it when we need to validate that both hostname override and NODE_NAME environment are not setted. 

The test case meta data should be 
```
map[string]struct {
		Opts         Options
                WantedNodeName string
	}
```

In each test, we should initialize `NodeProblemDetectorOptions` and base environment with `Opts` for such as setting NODE_NAME environment and hostname override. And, in the end of each test, we should check for the updated `NodeProblemDetectorOptions` instance and compare it with the desired `WantedNodeName`.